File: prisma/set-user-password.js
/* eslint-disable no-console */
/**
 * Reset a user's password (bcrypt 10 rounds, same as login API / admin reset).
 *
 * From project root (load .env — Node 20+: --env-file):
 *   node --env-file=.env prisma/set-user-password.js
 *   node --env-file=.env prisma/set-user-password.js --email user001@kpcqa.or.kr --password 2026000001
 *
 * Or npm (only if package.json "scripts" includes db:set-password — see below):
 *   npm run db:set-password -- --email user001@kpcqa.or.kr --password 2026000001
 *
 * If npm says "Missing script: db:set-password", add this next to db:seed in package.json:
 *   "db:set-password": "node prisma/set-user-password.js"
 *
 * Omit --password to use default 2026000001 (seed default for user001).
 * Add --no-must-change to skip forcing password change on next login.
 */
const { PrismaClient } = require("@prisma/client");
const bcrypt = require("bcryptjs");

const prisma = new PrismaClient();

function parseArgs(argv) {
  const out = {
    email: "user001@kpcqa.or.kr",
    password: "2026000001",
    mustChangePassword: true,
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--email" && argv[i + 1]) {
      out.email = String(argv[++i]).trim();
    } else if (a === "--password" && argv[i + 1]) {
      out.password = String(argv[++i]);
    } else if (a === "--no-must-change") {
      out.mustChangePassword = false;
    } else if (a === "-h" || a === "--help") {
      out.help = true;
    }
  }
  return out;
}

function printHelp() {
  console.log(`
Usage: node prisma/set-user-password.js [options]

Options:
  --email <addr>     User email (default: user001@kpcqa.or.kr)
  --password <pwd>   New password (default: 2026000001)
  --no-must-change   Do not set mustChangePassword=true

Examples:
  node --env-file=.env prisma/set-user-password.js
  node --env-file=.env prisma/set-user-password.js --email user001@kpcqa.or.kr --password 2026000001
`);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    printHelp();
    return;
  }

  const email = args.email.trim().toLowerCase();
  const password = String(args.password).trim();

  if (!email) {
    console.error("Error: email is required.");
    process.exit(1);
  }
  if (password.length < 6) {
    console.error("Error: password must be at least 6 characters.");
    process.exit(1);
  }

  const user = await prisma.user.findFirst({
    where: { email: { equals: email, mode: "insensitive" } },
    select: { id: true, email: true },
  });

  if (!user) {
    console.error(`Error: no user found for email: ${args.email}`);
    process.exit(2);
  }

  const passwordHash = await bcrypt.hash(password, 10);
  await prisma.user.update({
    where: { id: user.id },
    data: {
      passwordHash,
      mustChangePassword: args.mustChangePassword,
    },
  });

  console.log(
    `Updated password for ${user.email} (id=${user.id}, mustChangePassword=${args.mustChangePassword}).`
  );
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: scripts/set_user_password.py
"""
Set a user's password in the KPI Mgmt PostgreSQL DB (bcrypt, 10 rounds — same as bcryptjs in the app).

Prerequisites (once):
  pip install bcrypt psycopg2-binary python-dotenv

Usage (from project root, with .env containing DATABASE_URL):
  python scripts/set_user_password.py --email user001@kpcqa.or.kr --password 2026000001

Or pass the connection string:
  set DATABASE_URL=postgresql://...
  python scripts/set_user_password.py --email user001@kpcqa.or.kr
"""

from __future__ import annotations

import argparse
import os
import sys
from pathlib import Path


def load_env_file(path: Path) -> None:
    if not path.is_file():
        return
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, val = line.partition("=")
        key = key.strip()
        val = val.strip().strip('"').strip("'")
        if key and key not in os.environ:
            os.environ[key] = val


def main() -> int:
    root = Path(__file__).resolve().parents[1]
    load_env_file(root / ".env")

    try:
        import bcrypt
        import psycopg2
    except ImportError:
        print(
            "Missing packages. Run: pip install bcrypt psycopg2-binary python-dotenv",
            file=sys.stderr,
        )
        return 1

    parser = argparse.ArgumentParser(description="Reset User.passwordHash in KPI Mgmt DB")
    parser.add_argument(
        "--email",
        default="user001@kpcqa.or.kr",
        help="User email (case-insensitive match)",
    )
    parser.add_argument(
        "--password",
        default="2026000001",
        help="New plain password (default: 2026000001, matches seed employeeNo for user 001)",
    )
    parser.add_argument(
        "--no-must-change",
        action="store_true",
        help="Do not set mustChangePassword=true (default: force change on next login)",
    )
    args = parser.parse_args()

    database_url = os.environ.get("DATABASE_URL", "").strip()
    if not database_url:
        print("DATABASE_URL is not set (.env or environment).", file=sys.stderr)
        return 1

    if len(args.password) < 6:
        print("Password must be at least 6 characters (same rule as admin API).", file=sys.stderr)
        return 1

    password_hash = bcrypt.hashpw(args.password.encode("utf-8"), bcrypt.gensalt(rounds=10)).decode(
        "ascii"
    )
    must_change = not args.no_must_change

    conn = psycopg2.connect(database_url)
    try:
        with conn.cursor() as cur:
            cur.execute(
                """
                UPDATE "User"
                SET "passwordHash" = %s,
                    "mustChangePassword" = %s,
                    "updatedAt" = NOW()
                WHERE LOWER("email") = LOWER(%s)
                """,
                (password_hash, must_change, args.email.strip()),
            )
            if cur.rowcount == 0:
                print(f"No user found with email: {args.email}", file=sys.stderr)
                return 2
        conn.commit()
    finally:
        conn.close()

    print(f"Updated password for {args.email} (mustChangePassword={must_change}).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
